### PR TITLE
Pipe Dispenser UI tweaks

### PR DESCRIPTION
**Pull request by MTandi**
RPD:
g6iMANATn2

Pipe dispenser machines:
eBxH3wMlPu

GIF:
dreamseeker_LN0XizcPi9

## About The Pull Request

Updated RPD UI for:

- Bigger color buttons
- Faster selection of pipe type

## Why It's Good For The Game

Less clicks and missclicks when working with RPD

## Changelog

:cl:
qol: RPD and Pipe Pispenser UI tweaks for better navigation
code: RPD and Pipe Dispenser UI now on TS
/:cl:

### CODE REVIEW

**Comment by jlsnow301** [comment]
you know what I'm going to ask here: Can you convert this to typescript also? There are some other possible issues I see that existed previously in the ui, I'll note:

File: tgui/packages/tgui/interfaces/RapidPipeDispenser.js [hunk 1]
```diff
@@ -40,18 +40,19 @@ const TOOLS = [
...
   const { act, data } = useBackend(context);
-  const { space } = props;
   const { selected_color } = data;
   return (
     <LabeledList.Item label="Color">
-      {space ? <span>&nbsp;</span> : ''}
-      <Box inline width="64px" color={data.paint_colors[selected_color]}>
-        {capitalizeAll(selected_color)}
-      </Box>
       {Object.keys(data.paint_colors).map((colorName) => (
```
Comment (jlsnow301) [inline tgui/packages/tgui/interfaces/RapidPipeDispenser.js:46]: paint colors should be destructured and the object.keys array be made into a const prior to the jsx. lack of any type of optional chaining is making me a little nervous

**Review by jlsnow301 — COMMENTED** [review]
also here `const init_directions = data.init_directions || [];`
again, not your issues, but this could use the spruce up if you're willing,
`const { init_directions = [] } = data;`

File: tgui/packages/tgui/interfaces/RapidPipeDispenser.tsx [hunk 2]
```diff
@@ -38,59 +38,116 @@ const TOOLS = [
...
 ];
 
+type Directions = {
+  north: BooleanLike;
+  south: BooleanLike;
+  east: BooleanLike;
+  west: BooleanLike;
+};
+
+type Colors = {
```
Comment (jlsnow301) [inline tgui/packages/tgui/interfaces/RapidPipeDispenser.tsx:48]: are these being used as an enum? I'm confused what this is doing

Comment (MTandi) [reply tgui/packages/tgui/interfaces/RapidPipeDispenser.tsx:48]: I guess they're for forbidding certain connections for smart pipe. You can allow or forbid north connection for instance.

Comment (MTandi) [reply tgui/packages/tgui/interfaces/RapidPipeDispenser.tsx:48]: Do you want me to change this to a bitflag? I just wanted to make some tweaks to UI...

Comment (MTandi) [reply tgui/packages/tgui/interfaces/RapidPipeDispenser.tsx]: Renamed to DirectionsAllowed

Comment (jlsnow301) [reply tgui/packages/tgui/interfaces/RapidPipeDispenser.tsx]: that's totally fine. though a bitflag would be optimal, i think just a more descriptive name would be fine

File: code/game/objects/items/rcd/RPD.dm [hunk 3]
```diff
@@ -387,7 +380,11 @@ GLOBAL_LIST_INIT(transit_tube_recipes, list(
 				if(GLOB.objects_by_id_tag[CHAMBER_SENSOR_FROM_ID(initial(sensor.chamber_id))] != null)
 					continue
 
-			r += list(list("pipe_name" = info.name, "pipe_index" = i))
+			r += list(list(
+				"pipe_name" = info.name,
+				"pipe_index" = i,
+				"previews" = info.get_preview(p_dir, info == recipe)
+				))
```
Comment (san7890) [inline code/game/objects/items/rcd/RPD.dm:387]:
```suggestion
				"previews" = info.get_preview(p_dir, info == recipe),
			))
```

File: code/game/objects/items/rcd/RPD.dm [hunk 4]
```diff
@@ -137,22 +137,16 @@ GLOBAL_LIST_INIT(transit_tube_recipes, list(
...
 	var/list/rows = list()
-	var/list/row = list("previews" = list())
-	var/i = 0
 	for(var/dir in dirs)
 		var/numdir = text2num(dir)
 		var/flipped = ((dirtype == PIPE_TRIN_M) || (dirtype == PIPE_UNARY_FLIPPABLE)) && (ISDIAGONALDIR(numdir))
-		row["previews"] += list(list(
-			"selected" = dirtype == PIPE_ONEDIR ? TRUE : (numdir == selected_dir),
+		rows += list(list(
+			"selected" = selected && (!selected_dir ? FALSE : dirtype == PIPE_ONEDIR ? TRUE : (numdir == selected_dir)),
```
Comment (san7890) [inline code/game/objects/items/rcd/RPD.dm:144]:
```suggestion
			"selected" = selected && (!selected_dir ? FALSE : (dirtype == PIPE_ONEDIR ? TRUE : (numdir == selected_dir))),
```

this is way easier for me to read

**Review by san7890 — CHANGES_REQUESTED** [review]
nits

**Pull request merged**